master/buildbot/reporters/http.py

```diff
@@ -41,11 +41,28 @@ def checkConfig(self, *args, **kwargs):
                          (self.__class__,))
         if not isinstance(kwargs.get('builders'), (type(None), list)):
             config.error("builders must be a list or None")
+        # check that our callbacks are functions
+        if kwargs.get('startCB') and not hasattr(kwargs['startCB'], '__call__'):
+            config.error("start callback must be callable")
+        if kwargs.get('reviewCB') and not hasattr(kwargs['reviewCB'], '__call__'):
+            config.error("review callback must be callable")
+        if kwargs.get('summaryCB') and not hasattr(kwargs['summaryCB'], '__call__'):
+            config.error("summary callback must be callable")
 
     @defer.inlineCallbacks
-    def reconfigService(self, builders=None, **kwargs):
+    def reconfigService(self, builders=None,
+                        startCB=None, startArg=None,
+                        reviewCB=None, reviewArg=None,
+                        summaryCB=None, summaryArg=None,
+                        **kwargs):
         yield service.BuildbotService.reconfigService(self)
         self.builders = builders
+        self.startCB = startCB
+        self.startArg = startArg
+        self.reviewCB = reviewCB
```

Comment:
Preface - I want to get to the best possible design here. So I'm asking questions to work through it together :)

1. I'm not holding reviewCB as a gerrit thing - I'm viewing it as a builder configuration thing. If you create "pipelines" (or collections of related builders), then a reviewCB makes sense to understand how the entire collection behaved. Does that make sense?

2. This PR introduces the concept of callbacks as opt-in, which I believe will reduce long term code duplication. Different HTTP services may have different interfaces for what type of post data they expect (hipchat vs. stash is a great example). So, I believe we should enable that flexibility in the base class. Where does your understanding of "how this should be" differ?

3. I'm not following the recommendation - can you elaborate?